### PR TITLE
CIVDT-613: Update case route

### PR DESCRIPTION
**Pull request by jacemale-hmrc**

### CODE REVIEW

File: app/models/UpdateCaseError.scala [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+sealed trait UpdateCaseError extends Product with Serializable
+
+object UpdateCaseError {
+  case object InvalidCaseId extends UpdateCaseError
+  case object CaseAlreadyClosed extends UpdateCaseError
+  final case class UnexpectedError(message: String, status: Int) extends UpdateCaseError
+
+  def fromEisError(error: EisError): UpdateCaseError = {
+    error match {
+      case EisError.BackendError(_, _, "03- Invalid Case ID") => UpdateCaseError.InvalidCaseId
```
Comment (jacemale-hmrc) [inline app/models/UpdateCaseError.scala:30]: there is a space missing before the hyphen, but it's been like that in the docs :man_shrugging:

File: app/models/UpdateCaseError.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      case EisError.BackendError(_, _, "04 - Requested case already closed") => UpdateCaseError.CaseAlreadyClosed
+      case EisError.BackendError(_, status, message) => UpdateCaseError.UnexpectedError(message, status.toInt)
+      case EisError.UnexpectedError(status, reason) => UpdateCaseError.UnexpectedError(reason, status)
+    }
+  }
+
+  implicit val writes: Writes[UpdateCaseError] = Writes {
+    case UpdateCaseError.InvalidCaseId => Json.obj("errorCode" -> 1, "errorMessage" -> "Invalid case ID")
+    case UpdateCaseError.CaseAlreadyClosed => Json.obj("errorCode" -> 2, "errorMessage" -> "Requested case is already closed")
+    case UpdateCaseError.UnexpectedError(message, _) => Json.obj("errorCode" -> 3, "errorMessage" -> message)
```
Comment (jacemale-hmrc) [inline app/models/UpdateCaseError.scala:40]: this is what we'll return to the frontend in case of errors

Comment (ian-ralphs) [reply app/models/UpdateCaseError.scala:40]: I think what we get from EIS might be slightly different to this anyway so I suspect we will have to update when we get final API spec.

File: it/data/SampleData.scala [hunk 3]
```diff
@@ -299,4 +299,25 @@ trait SampleData {
     )
   )
 
+  val updateCaseJson: JsObject = Json.obj(
+    "caseId" -> "C18",
+    "additionalInfo" -> "Additional Info",
+    "supportingDocumentTypes" -> Json.arr(
```
Comment (ian-ralphs) [inline it/data/SampleData.scala:305]: we don't have supportDocTypes in the update case interface.

File: it/endpoint/UpdateCaseEndpointSpec.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      FileTransferStub.success()
+
+      val request: WSRequest = buildRequest("/update-case")
+
+      val response: WSResponse = await(request.post(updateCaseJson))
+
+      response.status shouldBe Status.OK
+
+    }
+
```
Comment (ian-ralphs) [inline it/endpoint/UpdateCaseEndpointSpec.scala:43]: can we have a test for returning a bad_request with the error code and message? I know this is missing from create case but no need to fix that as it is not this story.

File: test/connectors/httpParsers/EisHttpParsersSpec.scala [hunk 5]
```diff
@@ -82,22 +82,25 @@ class EisHttpParsersSpec extends SpecBase {
...
-  "Parsing a non-200 response" should {
+  "Parsing a 400 response" should {
 
     val headers: Map[String, Seq[String]] = Map("x-correlation-id" -> Seq(correlationId))
-    val body: JsObject = Json.obj()
-    val response = HttpResponse(Status.BAD_REQUEST, body, headers)
 
-    "return an error" in {
-      createCaseHttpParser.read("", "", response) mustBe Left(ErrorModel(Status.BAD_REQUEST, "Non-success response"))
+    "the response is a specific EIS error" in {
```
Comment (ian-ralphs) [inline test/connectors/httpParsers/EisHttpParsersSpec.scala:95]: can we have a test for the 04 error message too please. And maybe 06 just to show that parses to InternalServer error

**Review by ian-ralphs — CHANGES_REQUESTED** [review]

File: test/connectors/httpParsers/EisHttpParsersSpec.scala (hunk 5 above)
Comment (jacemale-hmrc) [reply test/connectors/httpParsers/EisHttpParsersSpec.scala:95]: this is already covered in the `UpdateCaseErrorSpec`

File: it/data/SampleData.scala (hunk 3 above)
Comment (jacemale-hmrc) [reply it/data/SampleData.scala:305]: removed it :+1:

File: it/endpoint/UpdateCaseEndpointSpec.scala (hunk 4 above)
Comment (jacemale-hmrc) [reply it/endpoint/UpdateCaseEndpointSpec.scala:54]: done

**Pull request merged**